### ISSUE TITLE
Is the limit on number of SNode the limit on total number of SNode created or only the number inside one SNode tree?

### ISSUE THREAD

**Issue by lin-hitonami** [taichi-dev/taichi#4065]
**Describe the bug**
It seems that we have a limit of number of SNode in the LLVM backend `taichi_max_num_snodes`, and this limit only limits the number of SNode inside a SNode tree now. However, in the [LLVM runtime](https://github.com/taichi-dev/taichi/blob/master/taichi/runtime/llvm/runtime.cpp#L560-L561), this should be a limit for the total number of SNode if I understand correctly. 

In the experiment using the program in #4023, when the total number of SNode exceeds the limit, the program continue to run correctly(?), with speed slower than before.

**Comment by k-ye** [comment]
Right... The members you've pointed out are for sparse SNodes. Ideally, we should have one (`element_lists`, `node_allocators`) per SNodeTree. Yet this wasn't a trivial fix.

**Comment by ifsheldon** [comment]
I don't understand why we need to limit the number of SNodes. I search for `taichi_max_num_snodes` and it seems that this constant is only used in array declarations. 

I don't know much about the Taichi cpp implementation, but a naive question is: Can't we replace these array of fixed size with a vector? then probably we can have "infinite" SNodes.

**Comment by k-ye** [comment]
It has something to do with the sparse runtime. Basically, each sparse SNode maps to an instance of some data structure in the sparse runtime. These instances are  linearly stored inside an array, and are indexed by the SNode id. Therefore Taichi  limits the max number of SNodes you can use, because it has to pre-allocate these arrays. 

One simple workaround is to just make these arrays dynamic. This would waste lots of space, because only the sparse SNodes use the aforementioned data structure. We can probably compact the array by giving each sparse SNode another id, e.g., `sparse_id`, and use this as the index in the array.
